Support skip

firelizzard18 opened the pull request:
Add support for `t.Skip()`, `t.Fail()`, and friends by implementing a custom `T`.

firelizzard18 committed:
Support skip

firelizzard18 committed:
tell libfuzzer not to use values that are too short

AdamKorcz commented:
Thank you for the PRs! Have all T methods been tested with libFuzzer?

AdamKorcz commented:
Could you add a few lines with a high level description about how support for `t.Skip` and friends is supported?

firelizzard18 commented:
>

firelizzard18 commented:
Wrong button...

firelizzard18 committed:
Tweaks and documentation

firelizzard18 commented:
I decided that failing the test should probably do something. If a test fails, that should probably be counted as a crasher. That's how the native fuzzing works. I wasn't sure if calling `os.Exit` was a good idea, so I added a panic to make sure libfuzzer would mark the input as a crasher.

I updated the documentation. I also changed `t.TempDir` to create a new tempdir.

firelizzard18 committed:
tests

firelizzard18 committed:
tests

firelizzard18 commented:
I added Go tests to verify the behavior of the supported T methods. I added test files that can be used to manually verify the behavior of skipping and failing.

AdamKorcz commented on utils/test_helpers.go:
```diff
@@ -1,27 +1,228 @@
+	<-done
+	return !s.failed
+}
+func (t *T) run(done chan<- struct{}, f func(t *T)) {
```
Has `T.run()` been tested with libFuzzer?

AdamKorcz commented:
> Thank you for the PRs! Have all T methods been tested with libFuzzer?

@firelizzard18 Could you comment on this please?

Edit: The tests help a lot! Thank you for those. For this PR to get merged (which would be great), all `T` methods need to be verified in libFuzzer to work as intended. 

To verify, we would need one, or several fuzzers, that include all T methods and that is/are then compiled and tested following these steps: https://github.com/AdamKorcz/go-118-fuzz-build#to-use. Given the size of the changes, it could be done in a Dockerfile that builds the fuzzers, and they can then be run and verified manually for now.

AdamKorcz commented on utils/test_helpers.go:
```diff
@@ -1,27 +1,228 @@
+func (t *T) Helper()      {}
+func (t *T) Name() string { return t.name }
+func (t *T) Parallel()    {}
+func (t *T) Run(name string, f func(t *T)) bool {
```
Do you recon we need well-defined `t.Run()` in go-118-fuzz-build? Will anyone ever use it inside `f.Fuzz()`?

firelizzard18 replied:
I use it. In some cases in my tests, slightly altering the input values provides better coverage. I run the unaltered inputs as the 'main' test and the altered inputs as a sub test.

AdamKorcz commented on utils/test_helpers.go:
```diff
@@ -1,27 +1,228 @@
+	t.mu.Lock()
+	t.finished = true
+	t.mu.Unlock()
+	runtime.Goexit()
```
Not sure this is the right approach for libFuzzer. I think `os.Exit()` would be better here.

firelizzard18 replied:
This is copy-pasted from `testing`. IMO the behavior of a binary built with go-118-fuzz-build should be as similar to the native fuzzing support as we can achieve with a reasonable effort. `runtime.Goexit` behaves similarly to a panic in that it calls deferred functions. Functions passed to `T.Cleanup()` are called by a deferred closure. If I replaced this with `os.Exit()`, cleanup functions would not be called if a test fails. Fuzz tests probably should not be doing things like creating temp files, but if someone does that I want `T.Cleanup()` to work the same as it does for native fuzzing.

AdamKorcz replied:
Has it been tested in a libFuzzer fuzzer?

AdamKorcz commented on utils/test_helpers.go:
```diff
@@ -1,27 +1,228 @@
+	defer t.mu.Unlock()
+	t.failed = true
+}
+func (t *T) FailNow() {
```
Can we log the stack trace when failing?

firelizzard18 replied:
I could print out the stack trace or save it and print it during cleanup. Do you have a preference?

AdamKorcz commented on utils/test_helpers.go:
```diff
@@ -1,27 +1,228 @@
+	recoverAndReturnPanic
+)
+
+func (t *T) log(s string) { _, _ = t.output.WriteString(s) }
```
For libFuzzer it is better to print out the string immediately instead of storing it in a string builder and print it out later. So could we do `fmt.Printlf(s)` instead of `t.output.WriteString(s)`?

favonia replied:
@AdamKorcz `Printf` seems wrong because `s` may contain things like `%s`. In any case, it seems we can just remove this helper function.

firelizzard18 commented:
> To verify, we would need one, or several fuzzers, that include all T methods and that is/are then compiled and tested following these steps: https://github.com/AdamKorcz/go-118-fuzz-build#to-use. Given the size of the changes, it could be done in a Dockerfile that builds the fuzzers, and they can then be run and verified manually for now.

@AdamKorcz Do you have a suggestion on what that would look like? I added a couple of test fuzzers in test data. I do not know how to verify that `Skip` results in `return -1` from the LLVM function. To verify that `Skip` stops execution, I could add `log.Fatal` or equivalent - is that the kind of thing you're thinking about?

AdamKorcz commented:
> > To verify, we would need one, or several fuzzers, that include all T methods and that is/are then compiled and tested following these steps: https://github.com/AdamKorcz/go-118-fuzz-build#to-use. Given the size of the changes, it could be done in a Dockerfile that builds the fuzzers, and they can then be run and verified manually for now.
> 
> @AdamKorcz Do you have a suggestion on what that would look like? I added a couple of test fuzzers in test data. I do not know how to verify that `Skip` results in `return -1` from the LLVM function. To verify that `Skip` stops execution, I could add `log.Fatal` or equivalent - is that the kind of thing you're thinking about?

It is great with the Go tests, but ultimately we will want this to work for libFuzzer. Could you let me know what in this PR has been tested with libFuzzer?

AdamKorcz commented:
> Do you have a suggestion on what that would look like?

A 1.18 fuzzer that has `Skip` and a `panic` after the `Skip`. We should be able to convert it to a libFuzzer fuzzer and run it without panicking.

favonia commented:
Hi, I am very interested in using `t.Failed` and other functions in my project, and this PR looks promising. :smiley: How can I help to move it forward?

firelizzard18 commented:
@favonia I have decided my purposes are better served by using my own fork, so I've pretty much abandoned this PR. You are welcome to fork my branch and submit your own PR based on my work.

AdamKorcz commented:
@favonia Some testing methods have been implemented here: https://github.com/AdamKorcz/go-118-fuzz-build/blob/main/testing/t.go. Feel free to add logic in there in a way that is meaningful for libFuzzer and preferably in small PRs

favonia commented:
@AdamKorcz I want clarifications on "meaningful for libFuzzer". Do you mean that new features should work as expected with libFuzzer, or that there is something else defining what's meaningful? And, is it correct that the tricky point is `os.Exit` v.s. `panic` v.s. `runtime.Goexit`? What else should we pay attention to?

favonia commented:
@AdamKorcz I will be occupied for the following months, so perhaps it is better to consider me unavailable. :sweat_smile:

AdamKorcz commented:
> @AdamKorcz I will be occupied for the following months, so perhaps it is better to consider me unavailable. sweat_smile

Alright. Feel free to open issues about suggestions or things that would improve G118FB for you.
